# Solutions for pythonchallenge.com puzzles

## [0. warming up](http://www.pythonchallenge.com/pc/def/0.html)
#### Problem:
This is a picture of 2**38.
#### Solution:
This is a simple intro. Just use pow() or 2**38.

In [71]:
2**38

274877906944

#### Answer:
274877906944

## [1. map](http://www.pythonchallenge.com/pc/def/map.html)

#### Problem:
Decode the messages.

#### Hints:
- The page title is 'What about making trans?'
- The page's image display a mapping:
    + K --> M
    + O --> Q
    + E --> G
    
#### Solution:
This is a basic Ceasar's cypher.

The hints suggests using string.maketrans().  This works, but it requires a translation table. It also seems to have issues in Python 3. string.maketrans was deprecated and then removed.  So in Python3, str.maketrans() is needed. Since Strings are not bytes, but Unicode a b'' prefix is needed.

I had considered using a dictionary LUT, but felt it makes reusability much harder. If the mapping K--> M changed to something else like K --> O, then a new table is required. One could write a function to generate the LUT or translation table for the required shift.   

Instead, string.ascii_lowercase and string.find() were used to get a character's index in the ascii string.  Then it becomes a matter of handling the wrapping. Positive/negative shifts and upper/lower case strings are handled.

In [72]:
%cd /home/ohlemacher/dev/pychallenge/1_think_twice
import think
think = reload(think)
msg_orig = \
        "g fmnc wms bgblr rpylqjyrc gr zw fylb. " + \
        "rfyrq ufyr amknsrcpq ypc dmp. bmgle gr gl zw fylb gq glcddgagclr" + \
        "ylb rfyr'q ufw rfgq rcvr gq qm jmle. sqgle qrpgle.kyicrpylq() gq pcamkkclbcb. lmu ynnjw ml rfc spj."
think.explore(msg_orig, 2)

/home/ohlemacher/dev/pychallenge/1_think_twice

Encrypted message:
g fmnc wms bgblr rpylqjyrc gr zw fylb. rfyrq ufyr amknsrcpq ypc dmp. bmgle gr gl zw fylb gq glcddgagclrylb rfyr'q ufw rfgq rcvr gq qm jmle. sqgle qrpgle.kyicrpylq() gq pcamkkclbcb. lmu ynnjw ml rfc spj.

Decrypted message:
i hope you didnt translate it by hand. thats what computers are for. doing it in by hand is inefficientand that's why this text is so long. using string.maketrans() is recommended. now apply on the url.


In [73]:
print think.shift_msg("Hap'o paop lkoepera wjz jacwpera odebpo.",  4)
print think.shift_msg("Let's test positive and negative shifts.", -4)

Let's test positive and negative shifts.
Hap'o paop lkoepera wjz jacwpera odebpo.


#### Answer:

In [74]:
print think.shift_msg("map", 2)

ocr


## [2. ocr](http://www.pythonchallenge.com/pc/def/ocr.html)
#### Problem:
A picture of a book open to some unreadable text.

#### Hints:
- The image has a caption 'recognize the characters. maybe they are in the book, but MAYBE they are in the page' source.
- The source contains 'find rare characters in the mess below:'
- The mess is a large block of text '%%$@$^#)^)&!_+]!*@&^}....'

#### Solution:
Count the characters in the text block. The answer is found in the characters that are used only once in the order they appear.

A dictionary `{char: count}` was used to collect the counts, then the dictionary was sorted: `sorted(dic.items(), key=operator.itemgetter(1))`. Fortunately, the built-in sorted() function is guaranteed to be stable. A sort is stable if it guarantees not to change the relative order of elements that compare equal.

#### Answer: equality

## [3. equality](http://www.pythonchallenge.com/pc/def/equality.html)

#### Problem:
An image of candles of various heights.

#### Hints:
- The page title is 'RE'.
- The image has a caption 'One small letter, surrounded by EXACTLY three big bodyguards on each of its sides.'
- The page source has a large block of text: 'kAewtloYgcFQaJNhHVG...'.

#### Solution:
Use a RE to find the matches:

`(^|[^A-Z]{1}) # Beginning or 3 non-uppercase
[A-Z]{3}      # Three uppercase (guard)
([a-z]{1})    # One lowercase
[A-Z]{3}      # Three uppercase (guard)
($|[^A-Z]{1}) # End or 3 non-uppercase`

#### Unit Testing:

In [80]:
%cd /home/ohlemacher/dev/pychallenge/3_equality
%run equality.py

....

/home/ohlemacher/dev/pychallenge/3_equality



----------------------------------------------------------------------
Ran 4 tests in 0.002s

OK


#### Answer:

In [76]:
%cd /home/ohlemacher/dev/pychallenge/3_equality
import equality
equality.explore()


/home/ohlemacher/dev/pychallenge/3_equality
linkedlist


## [4. linkedlist](http://www.pythonchallenge.com/pc/def/linkedlist.html)

#### Problem: 
Start at a url and follow the links until the end. There are a couple of special cases to handle with re.

#### Hints:
- follow the chain.
- urllib may help. DON'T TRY ALL NOTHINGS, since it will never end.
- 400 times is more than enough.
- The starting link is 12345.

#### Solution:
Use requests to traverse the links. Use re to find the next link.

#### Answer: peak

## [5. peak](http://www.pythonchallenge.com/pc/def/peak.html)

#### Problem:
The page source implies there is a source file to get. Figure out what to do with it.

#### Hints:
- peakhell src="banner.p"
- peakhell sounds a bit like pickle, but this is not very obvious.

#### Solution:
Use requests to get the banner.p file. Figure out it is a pickle file. This is the hard part. Demarshal the file using the pickle module. Print the text.  It contains lines of tuples (# or a space, multiplier), readable once printed by printing the character multiplier times.

#### Answer: channel

## [6. channel](http://www.pythonchallenge.com/pc/def/channel.html)

#### Problem:
Look at the picture and source and realize that a zip file must be downloaded and analyzed.

#### Hints:
- The picture is of a zipper.
- start from 90052.
- answer is inside the zip.

#### Solution:
Use requests to download channel.zip. Extract it and use recursion to follow each file in the zip file to the next. Collect each file's comments while following. Stop recursive calls once an interesting file is found. Print the comments. The comments are a banner: HOCKEY.

Another hint: 'it's in the air. look at the letters.'

The banner is formed from the banner characters.

#### Answer: oxygen

## [7. oxygen](http://www.pythonchallenge.com/pc/def/oxygen.html)

#### Problem:
A picture of a stream with a line of gray blocks varying in intensity.

#### Hints:
- The page title is 'smarty'.

#### Solution:
Get the image using requests. Use the mahotas image processing library to create a numpy ndarray with the pixel data. Maybe the gray blocks have some information? Find them. Being a level of gray, the pixels each have equal values for red, green and blue. The gray values map to ASCII characters. Use chr() to convert to a character.
- Note: The blocks each have many pixels, horizontally and vertically (it is a block). This leads to repeating strings and characters within the strings. Attempting to dedup can cause issues. Take care with gray values such as 110 which could be incorrectly deduped to 10.

#### Answer: integrity

## [8. integrity](http://www.pythonchallenge.com/pc/def/integrity.html)

#### Problem:
We have a picture of a bee with a hyperlink. When clicked, we a prompted for credentials, but we do not know how to log in.

#### Hints:
- The page source contains:
    - `un: 'BZh91AY&SYA\xaf\x82\r\x00\x00\x01\x01\x80\x02\xc0\x02\x00 \x00!\x9ah3M\x07<]\xc9\x14\xe1BA\x06\xbe\x084'`
    - `pw: 'BZh91AY&SY\x94$|\x0e\x00\x00\x00\x81\x00\x03$ \x00!\x9ah3M\x13<]\xc9\x14\xe1BBP\x91\xf08'`

#### Solution:
The un and pw strings are our credentials. They each start with 'BZ'. They are compressed with bzip compression. Use the bz2 module to decompress each.

#### Answer:
- user  : huge
- passwd: file
- Using the authentication leads to good.html

## [9. good](http://www.pythonchallenge.com/pc/return/good.html)

#### Problem:
We have a picture of a mountain view over water. There are some seemingly randomly placed black dots.

#### Hints:
- The page title is `Connect the dots`.
- The page source contains:
    - `first+second=?`
    - Two strings of numbers named first and second. They have different lengths.

#### Solution:
first and second are of different lengths so they cannot be coordinates used together. They instead each represent a line. The numbers are coordinates concatenated, e.g. [x0, y0, x1, y1...]. Use zip() to create two lists of coordinate tuples for both first and second. Use the PIL library to draw each line. The images's black dots are a red herring.

![alt text](https://raw.githubusercontent.com/ohlemacher/pychallenge/master/images/bull.png "bull")

#### Answer:
The lines form the outline of a cow but it has horns so it is really a 'bull'.

## [10. bull](http://www.pythonchallenge.com/pc/return/bull.html)

#### Problem:
After examining a sequence of integers, determine the length of the 30th iteration.

#### Hints:
- The page source contains `a = [1, 11, 21, 1211, 111221,`

#### Solution:
I had no idea what this sequence was and had to google it, though I had first guessed wrongly that the numbers were in base 3. Turns out this is the look-say sequence which was developed by John Conway.

Use regex to find the numbers that match "^(%s+)". Do this in a for loop sliding through the number. As you slide, skip the repeats. Create the next iteration by writing the length and the number as you slide.
        
#### Background of the look-say algorythm:
- [look-say backstory](https://www.youtube.com/watch?v=ea7lJkEhytA)
- "The stupidest problem you could conceivably imagine, that led to the most complicated answer you could conceivably imagine." -- John Conway
- The reason it is "the most complicated answer" is that Dr. Conway found a 70th order equation for this algorythm. Another solution route?
- Conway's constant: 1.303577269
- a[n+1] ~= a[n]**conways_constant

#### Answer: 5808

## [11. 5808](http://www.pythonchallenge.com/pc/return/5808.html)

#### Problem:
Given this image, find the solution.
![alt text](https://raw.githubusercontent.com/ohlemacher/pychallenge/master/images/5808.png "5808")

#### Hints:
- Page title: odd_even

#### Solution:
odd_even is pretty big hint. Remove the odd lines and rows. Using a numpy array makes it trivial. Once you have the ndarray via mahotas, the solution is one line: `img_even = img[::2, ::2]`
![alt text](https://raw.githubusercontent.com/ohlemacher/pychallenge/master/images/5808_evil.png "5808")

#### Answer: evil

## [12. evil](http://www.pythonchallenge.com/pc/return/evil.html)

#### Problem: Given the image named 'evil1.jpg', find the solution. The image is of someone dealing cards.

#### Hints:
- There is a one in the image name.
- Dealing cards is a small hint.
- On the back of each card are two sickle shapes.

#### Solution:
Since there is an evil1.jpg, try evil2.jpg. It exists and says 'not jpg - .gfx'

evil2.gfx is not a standard graphics format. The `file` command reports it as data. `xxd evil2.gfx | head` shows there is a jpeg magic number.

Remembering the solution to 5808, guess that there are images muxed together. Create individual data lists each skipping starting with a different offset.  After some experimentation, the offset is found to be five. Therefore we need five lists of bytes.

Look at the head of each list to determine its image format. Set the extension and create the files. Display each file. Catch errors. It turns out that image 3.png contains a compression error and results in a TypeError exception. Handle exceptions.  gimp can display the top half of 3.png. pylab cannot.

#### Answer: dis + pro + port + ional + !ity = disproportional

## [13. disproportional](http://www.pythonchallenge.com/pc/return/disproportional.html)

#### Problem:
This time we have an image of a phone dial pad.

#### Hints:
- The page title is 'call him'
- The image has a caption 'phone that evil'
- The page source has something interesting for this caption 'phone that <remote /> evil'
- The image's 5 button links to phonebook.php. This script returns an xml doc but there is an error.
    - `faultCode=105` 
    - `faultString=XML error: Invalid document end at line 1, column 1`
    
#### Solution:
In progress...